src/librustc_typeck/check/mod.rs

```diff
@@ -2683,8 +2683,14 @@ fn check_expr_with_unifier<'a, 'tcx, F>(fcx: &FnCtxt<'a, 'tcx>,
                                         unifier: F) where
     F: FnOnce(),
 {
-    debug!(">> typechecking: expr={} expected={}",
-           expr.repr(fcx.tcx()), expected.repr(fcx.tcx()));
+    if fcx.inh.node_types.borrow().get(&expr.id).is_some() {
```

Comment:
Without this, the RHS ended being type checked (`check_expr_*`) twice. And that causes problems when RHS is a generic enum, eg `Some(0) == None` didn't type check (RUST_LOG reported unfulfilled `_: Sized` obligations on `None`), and the error message was that `None` needed type annotations: `Some(0) == None::<i32>()`.

This was the least invasive solution that I could think of to avoid type cheking the RHS twice. If you have any other idea, let me know.